### ISSUE TITLE
Create a new model in merlin to work with Keras framework

### ISSUE THREAD

**Issue by ronanki** [CSTR-Edinburgh/merlin#35]
Make use of Keras framework to train the models and use it with merlin. 

Something, in the lines of "models/deep_rnn.py"

**Comment by ishank26** [comment]
Hi, 
I was trying to model LSTM using keras along the lines of merlin LSTM example. I bumped into one issue-

Does merlin uses "stateful" LSTM akin to Keras i.e cell states(activations) are propogated through successive batches, instead of resetting it for every batch. Similar approach was used by Alex Graves: Generating Sequences With Recurrent Neural Networks, where cell states are reset after every 100 epochs.

"Stateful" defined in [Keras Docs](https://keras.io/layers/recurrent/): 
_stateful: Boolean (default False). If True, the last state for each sample at index i in a batch will be used as initial state for the sample of index i in the following batch._

I went through merlin code, but I was unable to develop an intuition. It would be helpful if you can clarify it.
Thanks!

**Comment by bajibabu** [comment]
If I understand correctly  RNN/LSTM implementation in Merlin, it's not stateful as in Keras. Since we are not saving the previous cell states after each batch. In RNN/LSTM, the batch size is equal to utterance size  and after each utterance the cell-states will reset. Thus, if you want to implement the LSTM model in Keras with stateful then you have to modify the dataprovider script in merlin to format the data in fixed batch size (batch_size, n_in, n_out, input_length).

**Comment by bajibabu** [comment]
I think we need to modify the dataprovider script for sequential model training to gain the computational performance as mention in the https://github.com/CSTR-Edinburgh/merlin/issues/22, even if it's not for Keras.

**Comment by ishank26** [comment]
@bajibabu @ronanki Remembering cell states should increase performance in theory. Each batch will prime the activations, thus accounting for long term dependencies. As far as I understand, currently merlin is using a kind of truncated BPTT i.e BPTT over a single utterance as cell states are reset after each batch.

**Comment by bajibabu** [comment]
I agree on what you said. But it is good to do a small study on how much the performance is increased and how much computational time and memory are increased.

**Comment by ronanki** [comment]
@ishank26 As @bajibabu said, the issue #22 has to be implemented first before doing anything else. At the moment, the batch size for sequence modelling is 1 utterance and all utterances are shuffled before passing it to dataprovider script. This has to be changed -- w.r.t batch size (in terms of no. of utterances), the shuffle has to be done accordingly and then should be passed to dataprovider script.
